File: LeetCode/twopoint/ReverseVowelsOfAString.java
package twopoint;

public class ReverseVowelsOfAString {
    public String reverseVowels(String s) {
        char[] strs = s.toCharArray();

        int left = 0, right = strs.length - 1;

        while (left < right) {
            while (left < right && !isVowel(strs[left])) {
                left++;
            }

            while (left < right && !isVowel(strs[right])) {
                right--;
            }

            if (left < right) {
                char temp = strs[left];
                strs[left] = strs[right];
                strs[right] = temp;
                left++;
                right--;
            }
        }

        return new String(strs);
    }

    private boolean isVowel(char ch) {
        return "aeiouAEIOU".indexOf(ch) >= 0;
    }

    public static void main(String[] args) {
        ReverseVowelsOfAString solution = new ReverseVowelsOfAString();
        System.out.println(solution.reverseVowels("IceCreAm")); // Output: "AceCreIm"
        System.out.println(solution.reverseVowels("leetcode")); // Output: "leotcede"
    }
}

File: LeetCode_Python/twopoint/ReverseVowelsOfAString.py
class ReverseVowelsOfAString:
    def reverseVowels(self, s: str) -> str:
        """
        :type s: str
        :rtype: str
        """
        vowels = set("aeiouAEIOU")
        s = list(s)
        i, j = 0, len(s) - 1
        while i < j:
            if s[i] not in vowels:
                i += 1
            elif s[j] not in vowels:
                j -= 1
            else:
                s[i], s[j] = s[j], s[i]
                i += 1
                j -= 1
        return "".join(s)


# Time Complexity: O(n), where n is the length of the string
# Space Complexity: O(n), we are using a list to store the characters of the string

if __name__ == "__main__":
    solution = ReverseVowelsOfAString()

    print("Expected Output : holle")
    print("Actual Output : ", solution.reverseVowels("hello"))

    print("Expected Output : leotcede")
    print("Actual Output : ", solution.reverseVowels("leetcode"))
